### PR TITLE
Call Python 3.8 doClassCleanups

### LINKED ISSUE #8032: Class cleanups in Python 3.8+ are not called
https://docs.python.org/3/library/unittest.html#unittest.TestCase.addClassCleanup

Did not see `doClassCleanups` mentioned anywhere in this repo, which is the method that should be called. See https://github.com/python/cpython/blob/0f221d09cad46bee38d1b7a7822772df66c53028/Lib/unittest/suite.py#L175

Repo:
```
import unittest

def cleanup():
    assert False

class MyTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.addClassCleanup(cleanup)
    
    def test_one(self):
        pass
```

With unittest:
```
.E
======================================================================
ERROR: tearDownClass (test_demo.MyTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/petter/source/pytesttest/test_demo.py", line 4, in cleanup
    assert False
AssertionError

----------------------------------------------------------------------
Ran 1 test in 0.000s
```

With Pytest:
```
collected 1 item

test_demo.py .                                                                                                                                                                       [100%]

==================================================================================== 1 passed in 0.03s ====================================================================================
```

### PR DESCRIPTION
**Pull request by PetterS**
Fixes #8032 

Pytest already supports doCleanups; I added tests for this.

### CODE REVIEW

**Review by bluetech — CHANGES_REQUESTED** [review]
Thanks @PetterS, LGTM with a few comments. The extensive tests are nice.

I see there is also a new `addModuleCleanup` but that'd be more tricky to support since the code for `setUpModule`/`tearDownModule` is not in the unittest plugin for some reason.

File: src/_pytest/unittest.py [hunk 1]
```diff
@@ -99,26 +99,49 @@ def _inject_setup_teardown_fixtures(self, cls: type) -> None:
...
         )
         if class_fixture:
             cls.__pytest_class_setup = class_fixture  # type: ignore[attr-defined]
 
         method_fixture = _make_xunit_fixture(
-            cls, "setup_method", "teardown_method", scope="function", pass_self=True
+            cls,
+            "setup_method",
+            "teardown_method",
+            "<no cleanup function>",
```
[the remark this answers is not in the document]
Comment (PetterS) [reply src/_pytest/unittest.py:116]: I agree. Done.

File: src/_pytest/unittest.py [hunk 2]
```diff
@@ -99,26 +99,49 @@ def _inject_setup_teardown_fixtures(self, cls: type) -> None:
...
+    pass_self: bool,
 ):
     setup = getattr(obj, setup_name, None)
     teardown = getattr(obj, teardown_name, None)
     if setup is None and teardown is None:
         return None
 
+    cleanup = getattr(obj, cleanup_name, lambda *args: None)
+
+    def docleanup(self):
```
[the remark this answers is not in the document]
Comment (PetterS) [reply src/_pytest/unittest.py:139]: Done.

Linter would not let me assign a lambda to a variable so it looks like this now.

File: src/_pytest/unittest.py [hunk 3]
```diff
@@ -130,16 +153,23 @@ def fixture(self, request: FixtureRequest) -> Generator[None, None, None]:
...
-            if pass_self:
-                setup(self, request.function)
-            else:
-                setup()
+            try:
+                if pass_self:
+                    setup(self, request.function)
+                else:
+                    setup()
+            except Exception:
```
[the remark this answers is not in the document]
Comment (PetterS) [reply src/_pytest/unittest.py:161]: Done.

File: testing/test_unittest.py [hunk 4]
```diff
@@ -1260,3 +1260,168 @@ def test_plain_unittest_does_not_support_async(testdir):
...
+                self.assertEqual(MyTestCase.values, [1])
+    """
+    )
+    reprec = testdir.inline_run(testpath)
+    passed, skipped, failed = reprec.countoutcomes()
+    assert passed == 2
+
+
+@pytest.mark.skipif(
+    sys.version_info < (3, 1), reason="Feature introduced in Python 3.1"
```
[the remark this answers is not in the document]
Comment (PetterS) [reply testing/test_unittest.py:1350]: Done

File: testing/test_unittest.py [hunk 5]
```diff
@@ -1260,3 +1260,168 @@ def test_plain_unittest_does_not_support_async(testdir):
...
+        """
+        import unittest
+        class MyTestCase(unittest.TestCase):
+            values = []
+            @classmethod
+            def setUpClass(cls):
+                def cleanup():
+                    cls.values.append(1)
+                cls.addClassCleanup(cleanup)
+            def test_one(self):
```
[the remark this answers is not in the document]
Comment (PetterS) [reply testing/test_unittest.py:1279]: Done.

File: testing/test_unittest.py [hunk 6]
```diff
@@ -1260,3 +1260,168 @@ def test_plain_unittest_does_not_support_async(testdir):
...
+                pass
+        class Second(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (PetterS) [reply testing/test_unittest.py]: Done!

[the remark this answers is not in the document]
Comment (PetterS) [reply testing/test_unittest.py]: Done!

**Comment by PetterS** [comment]
I think this patch still can be improved. The `pass_self` branches for cleanup are not taken. I added those for consistency.

**Review by bluetech — APPROVED** [review]
Thanks, I think it's good to go. Will merge in a couple of days in case someone else also wants to review.

**Comment by PetterS** [comment]
Cheers! This will help our internal development!

**Comment by PetterS** [comment]
@bluetech Do you know when the next release of pytest will be?

**Comment by bluetech** [comment]
@PetterS Planned for this Saturday (see #8101)

**Comment by PetterS** [comment]
Cheers, thanks!

**Pull request merged**